components/core/tests/test-StreamingReader.cpp

```diff
@@ -0,0 +1,161 @@
+#include <thread>
+#include <vector>
+
+#include <Catch2/single_include/catch2/catch.hpp>
+#include <curl/curl.h>
+
+#include "../src/clp/ErrorCode.hpp"
+#include "../src/clp/FileReader.hpp"
+#include "../src/clp/ReaderInterface.hpp"
+#include "../src/clp/StreamingReader.hpp"
+
+namespace {
+constexpr char const* cTestUrl{
+        "https://raw.githubusercontent.com/y-scope/clp/main/components/core/tests/"
+        "test_network_reader_src/random.log"
+};
+
+constexpr size_t cDefaultReaderBufferSize{1024};
+
+[[nodiscard]] auto get_ref_file_abs_path() -> std::filesystem::path {
+    std::filesystem::path const file_path{__FILE__};
+    auto const test_root_path{file_path.parent_path()};
+    return test_root_path / "test_network_reader_src" / "random.log";
+}
+
+/**
+ * Reads the content of a given reader into a memory buffer.
+ * @param reader
+ * @param in_mem_buf
+ * @param reader_buffer_size The size of the buffer used to create a buffer for the underlying
+ * `read` operation.
+ */
+auto read_into_memory_buffer(
+        clp::ReaderInterface& reader,
+        std::vector<char>& in_mem_buf,
+        size_t reader_buffer_size = cDefaultReaderBufferSize
+) -> void {
+    in_mem_buf.clear();
+    // NOLINTNEXTLINE(cppcoreguidelines-avoid-c-arrays)
+    auto const read_buffer{std::make_unique<char[]>(reader_buffer_size)};
+    size_t num_bytes_read{};
+    bool has_more_content{true};
+    while (has_more_content) {
+        has_more_content = reader.read(read_buffer.get(), reader_buffer_size, num_bytes_read);
+        std::string_view const view{read_buffer.get(), num_bytes_read};
+        in_mem_buf.insert(in_mem_buf.cend(), view.cbegin(), view.cend());
+    }
+}
+
+/**
+ * Runs a function with a timeout.
+ * @tparam Func The function to run. A lambda function is expected, which takes no parameter.
+ * @param timeout Timeout in milliseconds.
+ * @param func Function to run.
+ * @return true if the function completes without timeout triggered, false otherwise.
+ */
+template <typename Func>
+[[nodiscard]] auto run_with_timeout(std::chrono::milliseconds timeout, Func&& func) -> bool {
+    auto future{std::async(std::launch::async, std::forward<Func>(func))};
+    return future.wait_for(timeout) == std::future_status::ready;
+}
+}  // namespace
+
+TEST_CASE("streaming_reader_basic", "[StreamingReader]") {
+    clp::FileReader ref_reader;
+    ref_reader.open(get_ref_file_abs_path().string());
+    std::vector<char> ref_data;
+    read_into_memory_buffer(ref_reader, ref_data);
+    ref_reader.close();
+
+    REQUIRE((clp::ErrorCode_Success == clp::StreamingReader::init()));
+    clp::StreamingReader reader(cTestUrl);
+    std::vector<char> streamed_data;
+    read_into_memory_buffer(reader, streamed_data);
+    auto const ret_code{reader.get_curl_return_code()};
+    REQUIRE(ret_code.has_value());
+    // NOLINTNEXTLINE(bugprone-unchecked-optional-access)
+    REQUIRE((CURLE_OK == ret_code.value()));
+    REQUIRE((streamed_data == ref_data));
```

Comment:
You should check the error before checking the download content. Otherwise if the test failed, it returns on the first failed assertion, and you won't know if it's curl error or data corruption caused by our implementation.